### PR TITLE
UCT: extended CQ moderation to RX cq

### PR DESCRIPTION
**Pull request by hoopoepg**
- added CQ moderation configuration to recv CQ
- minor code refactoring to avoid code duplication
- changed default moderation parameters

### CODE REVIEW

File: src/uct/ib/base/ib_iface.c [hunk 1]
```diff
@@ -82,13 +82,13 @@ ucs_config_field_t uct_ib_iface_config_table[] = {
    ucs_offsetof(uct_ib_iface_config_t, tx.cq_moderation), UCS_CONFIG_TYPE_UINT},
 
 #if HAVE_DECL_IBV_EXP_CQ_MODERATION
-  {"TX_CQ_MODERATION_COUNT", "-1",
+  {"CQ_EVENT_MODERATION_COUNT", "65535", /* MAX UINT16 */
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.c]: Description still mentions send WQEs

File: src/uct/ib/base/ib_iface.c [hunk 2]
```diff
@@ -479,6 +479,40 @@ static ucs_status_t uct_ib_iface_create_cq(uct_ib_iface_t *iface, int cq_length,
...
+            .moderation.cq_period = (uint16_t)(config->cq_moderation_period)
+        };
+        if (ibv_exp_modify_cq(cq, &cq_attr, IBV_EXP_CQ_MODERATION)) {
+            ucs_error("ibv_exp_modify_cq() failed: %m");
+            status = UCS_ERR_IO_ERROR;
+        }
+    }
+#endif /* HAVE_DECL_IBV_EXP_CQ_MODERATION */
+
+    return status;
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.c]: you could just return UCS_OK here, status is not needed

File: src/uct/ib/base/ib_iface.c [hunk 3]
```diff
@@ -479,6 +479,40 @@ static ucs_status_t uct_ib_iface_create_cq(uct_ib_iface_t *iface, int cq_length,
...
+        ucs_error("CQ moderation count is too high: %u", config->cq_moderation_count);
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    if (config->cq_moderation_period > UINT16_MAX) {
+        ucs_error("CQ moderation period is too high: %u", config->cq_moderation_period);
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    if (config->cq_moderation_count < UINT16_MAX ||
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.c]: code style, need to use () for each statement

File: src/uct/ib/base/ib_iface.c (hunk 2 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_iface.c]: status is used for compilation when HAVE_DECL_IBV_EXP_CQ_MODERATION macro is not defined

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4566/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_iface.c (hunk 1 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_iface.c]: fixed

File: src/uct/ib/base/ib_iface.c (hunk 3 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_iface.c]: fixed

File: src/uct/ib/base/ib_iface.c (hunk 2 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_iface.c]: fixed

**Comment by hoopoepg** [comment]
bot:ornl:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4567/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4572/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_iface.c [hunk 4]
```diff
@@ -82,13 +82,21 @@ ucs_config_field_t uct_ib_iface_config_table[] = {
    ucs_offsetof(uct_ib_iface_config_t, tx.cq_moderation), UCS_CONFIG_TYPE_UINT},
 
 #if HAVE_DECL_IBV_EXP_CQ_MODERATION
-  {"TX_CQ_MODERATION_COUNT", "-1",
+  {"TX_CQ_EVENT_MODERATION_COUNT", "65535", /* MAX UINT16 */
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: we need a shorter name.. 
how about TX_EVENT_MOD_COUNT, TX_EVENT_MOD_PERIOD?
also, why max_uint16?

File: src/uct/ib/base/ib_iface.c [hunk 5]
```diff
@@ -82,13 +82,21 @@ ucs_config_field_t uct_ib_iface_config_table[] = {
...
    "Number of send WQEs for which event is requested.",
-   ucs_offsetof(uct_ib_iface_config_t, tx.cq_moderation_count), UCS_CONFIG_TYPE_INT},
+   ucs_offsetof(uct_ib_iface_config_t, tx.cq_moderation_count), UCS_CONFIG_TYPE_UINT},
 
-  {"TX_CQ_MODERATION_PERIOD", "-1",
-   "Time period to generate event.",
-   ucs_offsetof(uct_ib_iface_config_t, tx.cq_moderation_period), UCS_CONFIG_TYPE_INT},
+  {"TX_CQ_EVENT_MODERATION_PERIOD", "65535", /* MAX UINT16 */
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: need to specify this parameter in time units and provide more info in the documentation

File: src/uct/ib/base/ib_iface.c (hunk 4 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_iface.c]: renamed.
max_int16 is used as default maximum value for moderation value (it operates by uint16 values), but it looks like not very gracefully, updated to process 0 as default value to disable moderation

File: src/uct/ib/base/ib_iface.c (hunk 5 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_iface.c]: actually there is no time units in docs, I assume it is usec (at least experiments demonstrated unit)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4587/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4586/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by hoopoepg** [comment]
bot:ornl:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4600/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4622/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by hoopoepg** [comment]
bot:ornl:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4624/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:ornl:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4626/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4627/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4629/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4630/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:ornl:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4633/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by hoopoepg** [comment]
bot:ornl:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4638/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4640/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:ornl:retest

**Comment by hoopoepg** [comment]
bot:ornl:retest

**Comment by yosefe** [comment]
@MattBBaker can you pls post ORNL failure?

**Comment by MattBBaker** [comment]
```
 CXX    uct/ib/gtest-test_cq_moderation.o
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/ib/test_cq_moderation.cc: In member function ‘void test_uct_cq_moderation::test_period(uct_iface*)’:
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/ib/test_cq_moderation.cc:128: error: expected primary-expression before ‘.’ token
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/ib/test_cq_moderation.cc:128: error: expected primary-expression before ‘.’ token
make[2]: *** [uct/ib/gtest-test_cq_moderation.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory `/var/lib/jenkins/jobs/UCX/workspace/build-tests/1505992191/test/gtest'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/var/lib/jenkins/jobs/UCX/workspace/build-tests/1505992191'
make: *** [all] Error 2
Build step 'Execute shell' marked build as failure
```

**Comment by hoopoepg** [comment]
fixed... really strange compilation issue. I seen same issue on MS compilers only (there are lack in C99 support)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4655/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4658/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4669/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_iface.c [hunk 6]
```diff
@@ -479,6 +488,49 @@ static ucs_status_t uct_ib_iface_create_cq(uct_ib_iface_t *iface, int cq_length,
     return status;
 }
 
+
+static ucs_status_t uct_ib_iface_set_moderation(struct ibv_cq *cq,
+                                                unsigned count, double period_sec)
+{
+#if HAVE_DECL_IBV_EXP_CQ_MODERATION
+    unsigned period = (unsigned)(period_sec * UCS_USEC_PER_SEC);
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: unsigned period_usec

File: src/uct/ib/base/ib_iface.c [hunk 7]
```diff
@@ -479,6 +488,49 @@ static ucs_status_t uct_ib_iface_create_cq(uct_ib_iface_t *iface, int cq_length,
...
+
+    if (count > UINT16_MAX) {
+        ucs_error("CQ moderation count is too high: %u", count);
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: pls write the maximal allowed value

File: src/uct/ib/base/ib_iface.c [hunk 8]
```diff
@@ -479,6 +488,49 @@ static ucs_status_t uct_ib_iface_create_cq(uct_ib_iface_t *iface, int cq_length,
...
+        return UCS_ERR_INVALID_PARAM;
+    } else if (count == 0) {
+        /* in case if count value is 0 (unchanged default value) - set it to maximum
+         * possible value */
+        count = UINT16_MAX;
+    }
+
+    if (period > UINT16_MAX) {
+        ucs_error("CQ moderation period is too high: %u", period);
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: pls write the maximal allowed value

File: src/uct/ib/base/ib_iface.c [hunk 9]
```diff
@@ -479,6 +488,49 @@ static ucs_status_t uct_ib_iface_create_cq(uct_ib_iface_t *iface, int cq_length,
...
+        return UCS_ERR_INVALID_PARAM;
+    } else if (period == 0) {
+        /* in case if count value is 0 (unchanged default value) - set it to maximum
+         * possible value, the same behavior as counter */
+        period = UINT16_MAX;
+    }
+
+    if ((count < UINT16_MAX) ||
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: no need to split line

File: test/gtest/uct/ib/test_cq_moderation.cc [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+#include <infiniband/verbs.h>
+
+#define TEST_PERIOD (3ul * UCS_USEC_PER_SEC) /* wait for 3 secs to get statistics */
+#define NSEC_PER_USEC (UCS_NSEC_PER_SEC / UCS_USEC_PER_SEC)
+
+#define MODERATION_PERIOD 1000 /* usecs */
+#define EVENT_LIMIT (10 * TEST_PERIOD / MODERATION_PERIOD / NSEC_PER_USEC)
+#define MAX_REPEATS 1000
+
+#define _STR(_arg) #_arg
```
Comment (yosefe) [inline test/gtest/uct/ib/test_cq_moderation.cc]: there is such macro in ucs/sys/preprocessor.h - UCS_PP_MAKE_STRING

File: test/gtest/uct/ib/test_cq_moderation.cc [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+class test_uct_cq_moderation : public uct_test {
+protected:
+
+    void init() {
+        uct_test::init();
+
+        if (!(GetParam()->tl_name == "rc" || GetParam()->tl_name == "rc_mlx5" ||
+              GetParam()->tl_name == "ud" || GetParam()->tl_name == "ud_mlx5" ||
+              GetParam()->tl_name == "dc" || GetParam()->tl_name == "dc_mlx5")) {
```
Comment (yosefe) [inline test/gtest/uct/ib/test_cq_moderation.cc]: don't understand, looks like all IB transports are enabled?

File: test/gtest/uct/ib/test_cq_moderation.cc [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+            /* dc_mlx5 is masked due to unknown issue - CQ moderation doesn't work on DC */
+            UCS_TEST_SKIP_R("unsupported");
+        }
+
+        if (RUNNING_ON_VALGRIND) {
+            UCS_TEST_SKIP_R("skipping on valgrind");
+        }
+
+        set_config("IB_TX_CQ_MODERATION=1");
+        if (!(GetParam()->tl_name == "ud" || GetParam()->tl_name == "ud_mlx5")) {
```
Comment (yosefe) [inline test/gtest/uct/ib/test_cq_moderation.cc]: better have a positive check (e.g if the tl is rc/dc/rc_mlx5/dc_mlx5)

File: test/gtest/uct/ib/test_cq_moderation.cc [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+            set_config("RC_FC_ENABLE=n");
+        }
+#if HAVE_DECL_IBV_EXP_CQ_MODERATION
+        set_config("IB_TX_EVENT_MOD_PERIOD=" TO_STR(MODERATION_PERIOD) "us");
```
Comment (yosefe) [inline test/gtest/uct/ib/test_cq_moderation.cc]: can use ucs:to_string() from test/gtest/common/test_helpers.h

File: src/uct/ib/base/ib_iface.c (hunk 6 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_iface.c]: fixed

File: src/uct/ib/base/ib_iface.c (hunk 7 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_iface.c]: done

File: src/uct/ib/base/ib_iface.c (hunk 8 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_iface.c]: done

File: src/uct/ib/base/ib_iface.c (hunk 9 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_iface.c]: fixed

File: test/gtest/uct/ib/test_cq_moderation.cc (hunk 10 above)
Comment (hoopoepg) [reply test/gtest/uct/ib/test_cq_moderation.cc]: fixed

File: test/gtest/uct/ib/test_cq_moderation.cc (hunk 11 above)
Comment (hoopoepg) [reply test/gtest/uct/ib/test_cq_moderation.cc]: artefact macro (some DC tests failed before... fixed)

File: test/gtest/uct/ib/test_cq_moderation.cc (hunk 12 above)
Comment (hoopoepg) [reply test/gtest/uct/ib/test_cq_moderation.cc]: done

File: test/gtest/uct/ib/test_cq_moderation.cc (hunk 13 above)
Comment (hoopoepg) [reply test/gtest/uct/ib/test_cq_moderation.cc]: stl containers will require data type conversion to std::string (or std::stream). macro string is shorter and easy to read

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4676/ for details (Mellanox internal link).

File: test/gtest/uct/ib/test_cq_moderation.cc (hunk 13 above)
Comment (evgeny-leksikov) [reply test/gtest/uct/ib/test_cq_moderation.cc]: @hoopoepg 
`set_config(std::string("IB_TX_EVENT_MOD_PERIOD=") + ucs::to_string(MODERATION_PERIOD) + "us");`
looks even more readable

Comment (hoopoepg) [reply test/gtest/uct/ib/test_cq_moderation.cc]: comparing with
```
set_config("IB_TX_EVENT_MOD_PERIOD=" MODERATION_STR);
```
?
are you sure? :)

Comment (evgeny-leksikov) [reply test/gtest/uct/ib/test_cq_moderation.cc]: For me - yes, using macro in C++ code is redundantly.

Comment (hoopoepg) [reply test/gtest/uct/ib/test_cq_moderation.cc]: done

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4678/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4679/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4683/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4686/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_iface.c [hunk 14]
```diff
@@ -82,13 +83,21 @@ ucs_config_field_t uct_ib_iface_config_table[] = {
...
 
-  {"TX_CQ_MODERATION_PERIOD", "-1",
-   "Time period to generate event.",
-   ucs_offsetof(uct_ib_iface_config_t, tx.cq_moderation_period), UCS_CONFIG_TYPE_INT},
+  {"TX_EVENT_MOD_PERIOD", "0us",
+   "Time period to generate send event (0 - disabled).",
+   ucs_offsetof(uct_ib_iface_config_t, tx.cq_moderation_period), UCS_CONFIG_TYPE_TIME},
+
+  {"RX_EVENT_MOD_COUNT", "0",
+   "Number of recv WQEs for which event is requested (0 - disabled).",
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: "Number of received messages for which an event would be generated (0 - disabled)."

File: src/uct/ib/base/ib_iface.c [hunk 15]
```diff
@@ -82,13 +83,21 @@ ucs_config_field_t uct_ib_iface_config_table[] = {
    ucs_offsetof(uct_ib_iface_config_t, tx.cq_moderation), UCS_CONFIG_TYPE_UINT},
 
 #if HAVE_DECL_IBV_EXP_CQ_MODERATION
-  {"TX_CQ_MODERATION_COUNT", "-1",
-   "Number of send WQEs for which event is requested.",
-   ucs_offsetof(uct_ib_iface_config_t, tx.cq_moderation_count), UCS_CONFIG_TYPE_INT},
+  {"TX_EVENT_MOD_COUNT", "0",
+   "Number of send WQEs for which event is requested (0 - disabled).",
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: "Number of send completions for which an event would be generated (0 - disabled)."

File: src/uct/ib/base/ib_iface.c [hunk 16]
```diff
@@ -82,13 +83,21 @@ ucs_config_field_t uct_ib_iface_config_table[] = {
...
+  {"TX_EVENT_MOD_PERIOD", "0us",
+   "Time period to generate send event (0 - disabled).",
+   ucs_offsetof(uct_ib_iface_config_t, tx.cq_moderation_period), UCS_CONFIG_TYPE_TIME},
+
+  {"RX_EVENT_MOD_COUNT", "0",
+   "Number of recv WQEs for which event is requested (0 - disabled).",
+   ucs_offsetof(uct_ib_iface_config_t, rx.cq_moderation_count), UCS_CONFIG_TYPE_UINT},
+
+  {"RX_EVENT_MOD_PERIOD", "0us",
+   "Time period to generate recv event (0 - disabled).",
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: recv -> receive

File: src/uct/ib/base/ib_iface.c [hunk 17]
```diff
@@ -479,6 +488,48 @@ static ucs_status_t uct_ib_iface_create_cq(uct_ib_iface_t *iface, int cq_length,
...
+
+    if ((count < UINT16_MAX) || (period < UINT16_MAX)) {
+        struct ibv_exp_cq_attr cq_attr = {
+            .comp_mask            = IBV_EXP_CQ_ATTR_MODERATION,
+            .moderation.cq_count  = (uint16_t)(count),
+            .moderation.cq_period = (uint16_t)(period),
+            .cq_cap_flags         = 0
+        };
+        if (ibv_exp_modify_cq(cq, &cq_attr, IBV_EXP_CQ_MODERATION)) {
+            ucs_error("ibv_exp_modify_cq() failed: %m");
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: "ibv_exp_modify_cq(count=%d, period=%d) failed: %m"

File: test/gtest/uct/ib/test_cq_moderation.cc (hunk 11 above)
Comment (yosefe) [reply test/gtest/uct/ib/test_cq_moderation.cc]: so let's remove?

File: test/gtest/uct/ib/test_cq_moderation.cc [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+extern "C" {
+#include <uct/api/uct.h>
+}
+#include <uct/uct_test.h>
+#include <ucs/time/time.h>
+#include <poll.h>
+#include <infiniband/verbs.h>
+
+#define TEST_PERIOD (3ul * UCS_USEC_PER_SEC) /* wait for 3 secs to get statistics */
+#define NSEC_PER_USEC (UCS_NSEC_PER_SEC / UCS_USEC_PER_SEC)
```
Comment (yosefe) [inline test/gtest/uct/ib/test_cq_moderation.cc]: there is UCS_NSEC_PER_SEC in ucs/time/time.h

File: test/gtest/uct/ib/test_cq_moderation.cc [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+
+#define MODERATION_PERIOD 1000 /* usecs */
+
+#define EVENT_LIMIT (20 * TEST_PERIOD / MODERATION_PERIOD / NSEC_PER_USEC)
```
Comment (yosefe) [inline test/gtest/uct/ib/test_cq_moderation.cc]: let's define all those as C++ constants in the test class - e.g "static const unsigned moderation_period = 1000;"

File: test/gtest/uct/ib/test_cq_moderation.cc [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+        if (RUNNING_ON_VALGRIND) {
+            UCS_TEST_SKIP_R("skipping on valgrind");
+        }
+
+        set_config("IB_TX_CQ_MODERATION=1");
+        if ((GetParam()->tl_name == "rc") || (GetParam()->tl_name == "rc_mlx5") ||
+            (GetParam()->tl_name == "dc") || (GetParam()->tl_name == "dc_mlx5")) {
+            set_config("RC_FC_ENABLE=n");
+        }
+#if HAVE_DECL_IBV_EXP_CQ_MODERATION
```
Comment (yosefe) [inline test/gtest/uct/ib/test_cq_moderation.cc]: you can put just UCT_INSTANTIATE_NO_SELF_TEST_CASE inside the #if
then, will also not be needed to skip the test if unsupported

File: test/gtest/uct/ib/test_cq_moderation.cc [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+        do {
+            /* arm all event types */
+            while (1) {
+                if (uct_iface_event_arm(iface, UCT_EVENT_SEND_COMP | UCT_EVENT_RECV_AM | UCT_EVENT_RECV_SIG_AM) != UCS_ERR_BUSY) {
+                    break;
+                }
+                progress();
+            }
+            /* repeat till FD is in active state */
+        } while(poll(&pfd, 1, 0) > 0);
```
Comment (yosefe) [inline test/gtest/uct/ib/test_cq_moderation.cc]: code style: space after 'while' word

File: test/gtest/uct/ib/test_cq_moderation.cc [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+            polled = poll(&pfd, 1, 0);
+            if (polled > 0) {
+                events++;
+                iface_arm(iface);
+            }
+
+            do {
+                status = uct_ep_am_short(m_sender->ep(0), 0, 0, NULL, 0);
+                progress();
+            } while(status == UCS_ERR_NO_RESOURCE);
```
Comment (yosefe) [inline test/gtest/uct/ib/test_cq_moderation.cc]: code style: space after 'while' word

File: test/gtest/uct/ib/test_cq_moderation.cc [hunk 23]
```diff
@@ -0,0 +1,179 @@
...
+
+        uct_iface_event_fd_get(iface, &fd);
+
+        pfd.fd = fd;
+        pfd.events = POLLIN;
+
+        do {
+            /* arm all event types */
+            while (1) {
+                if (uct_iface_event_arm(iface, UCT_EVENT_SEND_COMP | UCT_EVENT_RECV_AM | UCT_EVENT_RECV_SIG_AM) != UCS_ERR_BUSY) {
```
Comment (yosefe) [inline test/gtest/uct/ib/test_cq_moderation.cc]: need to split long line?

File: test/gtest/uct/ib/test_cq_moderation.cc (hunk 18 above)
Comment (hoopoepg) [reply test/gtest/uct/ib/test_cq_moderation.cc]: right, but here is value for number nsecs in usecs (micro / nano)

Comment (yosefe) [reply test/gtest/uct/ib/test_cq_moderation.cc]: right, sorry.

File: test/gtest/uct/ib/test_cq_moderation.cc (hunk 11 above)
Comment (hoopoepg) [reply test/gtest/uct/ib/test_cq_moderation.cc]: comment is removed.
or you mean remove this evaluation for transport name?

File: src/uct/ib/base/ib_iface.c (hunk 15 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_iface.c]: done

**Comment by hoopoepg** [comment]
all requested changes are done

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4693/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4698/ for details (Mellanox internal link).

File: test/gtest/uct/ib/test_cq_moderation.cc (hunk 11 above)
Comment (yosefe) [reply test/gtest/uct/ib/test_cq_moderation.cc]: yes, remove the transport name check

Comment (hoopoepg) [reply test/gtest/uct/ib/test_cq_moderation.cc]: ok, got your point.
updated transport check: used _IB_ macro to test class instantiate

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4707/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
:+1: @hoopoepg can you pls squash?

**Comment by hoopoepg** [comment]
done: 2 commits: moderation + test

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4711/ for details (Mellanox internal link).

**Pull request merged**